File: languages/protocol/go/worker.go
package protocol

import (
	"bufio"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
)

const Version = "2.0.0"

func Arg(name string) string {
	args := os.Args
	for i := 0; i+1 < len(args); i++ {
		if args[i] == name {
			return args[i+1]
		}
	}
	panic(fmt.Sprintf("missing %s", name))
}

func DigestBytes(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

func EmitLine(value any) {
	encoded, err := json.Marshal(value)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(encoded))
}

// RunWorker executes the harness loop. kernel must return JSON-marshalable output.
func RunWorker(inputPath, outputPath string, kernel func() any) {
	if Arg("--protocol-version") != Version {
		panic("unsupported protocol version")
	}
	if _, err := os.ReadFile(inputPath); err != nil {
		panic(err)
	}

	EmitLine(map[string]string{"type": "ready", "protocolVersion": Version})

	var lastOutput []byte
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		var message map[string]any
		if err := json.Unmarshal([]byte(line), &message); err != nil {
			panic(err)
		}
		switch message["type"] {
		case "run":
			lastOutput, _ = json.Marshal(kernel())
			requestID, _ := message["requestId"].(float64)
			EmitLine(map[string]any{
				"type":      "result",
				"requestId": int64(requestID),
				"digest":    DigestBytes(lastOutput),
			})
		case "finish":
			if err := os.WriteFile(outputPath, lastOutput, 0o644); err != nil {
				panic(err)
			}
			EmitLine(map[string]string{"type": "finish", "digest": DigestBytes(lastOutput)})
			return
		default:
			panic(fmt.Sprintf("unknown protocol message type %v", message["type"]))
		}
	}
	if err := scanner.Err(); err != nil {
		panic(err)
	}
}

File: languages/protocol/worker.mjs
import { createHash } from "node:crypto";
import { createInterface } from "node:readline";
import { readFile, writeFile } from "node:fs/promises";

export const PROTOCOL_VERSION = "2.0.0";

export function arg(name, argv = process.argv) {
  const index = argv.indexOf(name);
  if (index < 0 || index + 1 >= argv.length) throw new Error(`missing ${name}`);
  return argv[index + 1];
}

export function digestBytes(bytes) {
  return createHash("sha256").update(bytes).digest("hex");
}

export function digestJson(value) {
  return digestBytes(Buffer.from(JSON.stringify(value)));
}

export function emitLine(value) {
  const line = typeof value === "string" ? value : JSON.stringify(value);
  process.stdout.write(`${line}\n`);
}

/**
 * Run the harness protocol loop.
 * @param {{ inputPath: string, outputPath: string, kernel: () => unknown }} options
 */
export async function runWorker({ inputPath, outputPath, kernel }) {
  if (arg("--protocol-version") !== PROTOCOL_VERSION) {
    throw new Error(`unsupported protocol version ${arg("--protocol-version")}`);
  }

  await readFile(inputPath, "utf8");
  emitLine({ type: "ready", protocolVersion: PROTOCOL_VERSION });

  const reader = createInterface({ input: process.stdin });
  let lastBytes = Buffer.alloc(0);

  for await (const line of reader) {
    if (!line) continue;
    const message = JSON.parse(line);
    if (message.type === "run") {
      lastBytes = Buffer.from(JSON.stringify(kernel()));
      emitLine({ type: "result", requestId: message.requestId, digest: digestBytes(lastBytes) });
    } else if (message.type === "finish") {
      await writeFile(outputPath, lastBytes);
      emitLine({ type: "finish", digest: digestBytes(lastBytes) });
      break;
    } else {
      throw new Error(`unknown protocol message type ${message.type}`);
    }
  }
}
